### DISCUSSION TITLE
Asynchronous scoped sessions not torn down and running in different loops.

### DISCUSSION THREAD

**Discussion by EM51641** [sqlalchemy/sqlalchemy#10384] [Usage Questions]
Hello,

First, I want to thank you for this fantastic library. When I try to run my "asynchronous" app on Flask, I get that the session is attached to a different loop. However, the controller get executed. It is at the context teardown level that an issue happens, but I do not understand how.

The error I get is:

```python
RuntimeError: Task <Task pending name='Task-8' coro=<AsyncToSync.main_wrap() running at /usr/local/lib/python3.11/site-packages/asgiref/sync.py:353> cb=[_run_until_complete_cb() at /usr/local/lib/python3.11/asyncio/base_events.py:180]> got Future <Future pending cb=[Protocol._on_waiter_completed()]> attached to a different loop
```

I added a reproducible example using devcontainer at: https://github.com/EM51641/async_flask_sqlalchemy_app/tree/main. Could somebody please tell me what am I doing so wrong ?

**Comment by CaselIT** [comment]
Hi,

Not sure, you seem to somehow have two loops running. Asyncio is very rigid about the fact that you cannot mix instances from tho different loops. As far as I know this is the first time something similar was reported, so it may be worth asking also to Flask to see if the are using two loops for some reason?

Looking at the code quickly, maybe you can change the last two lines:
```py
    flask_manager = asyncio.run(create_app_manager())
    flask_manager.app.run(debug=True, threaded=False)
```
to instead be
```py
async def go():
    flask_manager = await create_app_manager()
    flask_manager.app.run(debug=True, threaded=False)
asyncio.run(go())
```

But I'm not sure if this will improve things or not

**Reply by EM51641** [reply]
Thank you for your answer. Unfortunately, I tried without success. I found out that the scoped method ```_get_current_context``` is being ran on another loop when I initialise the app, which is weird. I raised the point there: https://github.com/pallets/flask/discussions/5265

**Comment by EM51641** [comment]
Finally after 3 days of headaches I found that the problem came from the asyncpg driver, If I replace it by psycopg everything works fine.

**Reply by CaselIT** [reply]
That's strange! May be worth reporting it to them. Maybe there is something in it or in the way sqlalchemy uses it that's causing some issue?

**Reply by EM51641** [reply]
I frankly don't have any idea. Maybe the asyncpg configuration/implementation in sqlalchemy is problematic.
